File: src/main/scala/com/exasol/cloudetl/orc/converter/OrcConverterFactory.scala
package com.exasol.cloudetl.orc.converter

import com.exasol.errorreporting.ExaError

import org.apache.hadoop.hive.ql.exec.vector.ColumnVector
import org.apache.orc.TypeDescription
import org.apache.orc.TypeDescription.Category

/**
 * An Orc data types converter factory class.
 */
object OrcConverterFactory {

  /**
   * Given the Orc [[org.apache.orc.TypeDescription$]] types creates a
   * converter that reads the type value into Java objects.
   */
  def apply(orcType: TypeDescription): OrcConverter[_ <: ColumnVector] =
    if (orcType.getCategory().isPrimitive()) {
      createPrimitiveConverter(orcType)
    } else {
      createComplexConverter(orcType)
    }

  private[this] def createPrimitiveConverter(orcType: TypeDescription): OrcConverter[_ <: ColumnVector] =
    orcType.getCategory() match {
      case Category.BOOLEAN   => BooleanConverter
      case Category.BYTE      => ByteConverter
      case Category.SHORT     => ShortConverter
      case Category.INT       => IntConverter
      case Category.LONG      => LongConverter
      case Category.FLOAT     => FloatConverter
      case Category.DOUBLE    => DoubleConverter
      case Category.DECIMAL   => DecimalConverter
      case Category.DATE      => DateConverter
      case Category.TIMESTAMP => TimestampConverter
      case Category.BINARY    => BinaryConverter
      case Category.CHAR      => StringConverter
      case Category.STRING    => StringConverter
      case Category.VARCHAR   => StringConverter
      case _ =>
        throw new IllegalArgumentException(
          ExaError
            .messageBuilder("F-CSE-10")
            .message("Orc primitive type {{PRIMITIVE_TYPE}} is not supported.")
            .parameter("PRIMITIVE_TYPE", String.valueOf(orcType.getCategory()))
            .ticketMitigation()
            .toString()
        )
    }

  private[this] def createComplexConverter(orcType: TypeDescription): OrcConverter[_ <: ColumnVector] =
    orcType.getCategory() match {
      case Category.LIST =>
        val listElementType = orcType.getChildren().get(0)
        ListConverter(apply(listElementType))
      case Category.MAP =>
        val mapKeyType = orcType.getChildren().get(0)
        val mapValueType = orcType.getChildren().get(1)
        MapConverter(apply(mapKeyType), apply(mapValueType))
      case Category.STRUCT => StructConverter(orcType)
      case Category.UNION  => UnionConverter(orcType)
      case _ =>
        throw new IllegalArgumentException(
          ExaError
            .messageBuilder("F-CSE-11")
            .message("Orc complex type {{COMPLEX_TYPE}} is not supported.")
            .parameter("COMPLEX_TYPE", String.valueOf(orcType.getCategory()))
            .ticketMitigation()
            .toString()
        )
    }

}

File: src/main/java/com/exasol/cloudetl/orc/converter/OrcConverterFactory.java
package com.exasol.cloudetl.orc.converter;

import org.apache.hadoop.hive.ql.exec.vector.ColumnVector;
import org.apache.orc.TypeDescription;
import org.apache.orc.TypeDescription.Category;

import com.exasol.errorreporting.ExaError;

/** Factory for ORC converters. */
public final class OrcConverterFactory {
    private OrcConverterFactory() {
        // Utility class.
    }

    /** Create a converter for an ORC type. */
    public static OrcConverter<? extends ColumnVector> create(final TypeDescription orcType) {
        if (orcType.getCategory().isPrimitive()) {
            return createPrimitiveConverter(orcType);
        }
        return createComplexConverter(orcType);
    }

    /** Factory alias for Java callers. */
    public static OrcConverter<? extends ColumnVector> apply(final TypeDescription orcType) {
        return create(orcType);
    }

    private static OrcConverter<? extends ColumnVector> createPrimitiveConverter(final TypeDescription orcType) {
        switch (orcType.getCategory()) {
        case BOOLEAN:
            return OrcConverters.BOOLEAN;
        case BYTE:
            return OrcConverters.BYTE;
        case SHORT:
            return OrcConverters.SHORT;
        case INT:
            return OrcConverters.INT;
        case LONG:
            return OrcConverters.LONG;
        case FLOAT:
            return OrcConverters.FLOAT;
        case DOUBLE:
            return OrcConverters.DOUBLE;
        case DECIMAL:
            return OrcConverters.DECIMAL;
        case DATE:
            return OrcConverters.DATE;
        case TIMESTAMP:
            return OrcConverters.TIMESTAMP;
        case BINARY:
            return OrcConverters.BINARY;
        case CHAR:
        case STRING:
        case VARCHAR:
            return OrcConverters.STRING;
        default:
            throw new IllegalArgumentException(ExaError.messageBuilder("F-CSE-10")
                    .message("Orc primitive type {{PRIMITIVE_TYPE}} is not supported.")
                    .parameter("PRIMITIVE_TYPE", String.valueOf(orcType.getCategory())).ticketMitigation().toString());
        }
    }

    private static OrcConverter<? extends ColumnVector> createComplexConverter(final TypeDescription orcType) {
        final Category category = orcType.getCategory();
        switch (category) {
        case LIST:
            return new OrcConverters.ListConverter<>(create(orcType.getChildren().get(0)));
        case MAP:
            return new OrcConverters.MapConverter<>(create(orcType.getChildren().get(0)),
                    create(orcType.getChildren().get(1)));
        case STRUCT:
            return new StructConverter(orcType);
        case UNION:
            return new UnionConverter(orcType);
        default:
            throw new IllegalArgumentException(ExaError.messageBuilder("F-CSE-11")
                    .message("Orc complex type {{COMPLEX_TYPE}} is not supported.")
                    .parameter("COMPLEX_TYPE", String.valueOf(orcType.getCategory())).ticketMitigation().toString());
        }
    }
}
